### PR TITLE
First pass cleanup of bls_verify.md

### PR DESCRIPTION
**Pull request by JustinDrake**
Misc cleanups:

* (typo) `highflag` => `highflag1`
* (typo) `lowflag = x = 0` => `lowflag == x == 0`
* Add structure and table of contents
* Describe more notation in words (e.g. `i`)
* Make sure flags are 1-bit
* Clarify and polish presentation

Example notation cleanups:

* `G1` => `g` (to avoid confusing with the group G1)
* `field_modulus` => `q` (avoid using two names for same thing)
* `BLSVerify` => `bls_verify` (respect notation for functions in main document)
* `sig` => `signature` (avoid abbreviations as in main document)

TODO:

* Define argument and returned value types for `hash_to_G2`, `modular_squareroot`
* Describe/define `FQ2`, `b2`, `is_on_curve`, `multiply`, `g`
* Clarify where the value for `G2_cofactor` came from
* Make the naming changes around `bls_verify` in the main document
* Fix any bugs introduced by the cleanup

### CODE REVIEW

**Comment by JustinDrake** [comment]
cc @mratsim

File: specs/bls_verify.md [hunk 1]
```diff
@@ -1,64 +1,117 @@
...
-field_modulus = 4002409555221667393417789825735904156556882819939007885332058136124031650490837864442687629129015664037894272559787
+q = 4002409555221667393417789825735904156556882819939007885332058136124031650490837864442687629129015664037894272559787
 
-def hash_to_G2(m, domain):
-    x1 = hash(bytes8(domain) + b'\x01' + m)
-    x2 = hash(bytes8(domain) + b'\x02' + m)
-    x_coord = FQ2([x1, x2]) # x1 + x2 * i
+def hash_to_G2(message, domain):
+    x1 = hash(bytes8(domain) + b'\x01' + message)
+    x2 = hash(bytes8(domain) + b'\x02' + message)
```
Comment (hwwhww) [inline specs/bls_verify.md]:
```suggestion
    x2 = int.from_bytes(hash(bytes8(domain) + b'\x02' + message))
```

File: specs/bls_verify.md [hunk 2]
```diff
@@ -1,64 +1,117 @@
...
+### G1 points
 
-Here is the `hash_to_G2` definition:
+A point in G1 is represented as a 384-bit integer `z` decomposed as a 381-bit integer and three 1-bit flags:
+
+* `x = z % 2**381`
+* `a_flag = (z % 2**382) // 2**381`
+* `b_flag = (z % 2**383) // 2**382`
+* `c_flag = (z % 2**384) // 2**383`
+
```
Comment (ChihChengLiang) [inline specs/bls_verify.md]: Would be nice to 
- emphasize flags are prefixed. It takes a reader a while to figure out the relative position of bits from bullet points. `z = |c_flag|b_flag|a_flag|...x...|`
- and intepret the intention of these flags. Say, b_flag indicates if it is infinity, c_flag is dummy, while a_flag identifies which of two possible y this point has.

Comment (JustinDrake) [reply specs/bls_verify.md]: > emphasize flags are prefixed. It takes a reader a while to figure out the relative position of bits from bullet points. `z = |c_flag|b_flag|a_flag|...x...|`

Done :)

> and intepret the intention of these flags

The problem is that these flags have different meanings in G1 and G2.

File: specs/bls_verify.md (hunk 1 above)
Comment (JustinDrake) [reply specs/bls_verify.md]: Done (adding `, 'big'`)

File: specs/bls_verify.md [hunk 3]
```diff
@@ -1,64 +1,113 @@
...
-        if y_coord is not None:
+        x_cubed_plus_b2 = x_coordinate ** 3 + FQ2([4, 4])
+        y_coordinate = modular_squareroot(x_cubed_plus_b2)
+        if y_coordinate is not None:
             break
-        x_coord += FQ2([1, 0]) # Add one until we get a quadratic residue
-    assert is_on_curve((x_coord, y_coord))
-    return multiply((x_coord, y_coord), G2_cofactor)
+        x_coordinate += FQ2([1, 0]) # Add one until we get a quadratic residue
+    assert is_on_curve((x_coordinate, y_coordinate))
```
[the remark this answers is not in the document]
Comment (JustinDrake) [reply specs/bls_verify.md]: I'm tempted to leave it as is for consistency with `multiply_in_G2((x_coordinate, y_coordinate), G2_cofactor)`. Also arguably the point is `(x_coordinate, y_coordinate)` which is fed as an argument to `is_on_G2`.

**Comment by JustinDrake** [comment]
@vbuterin:

1) What is the role of `G2_cofactor` and where did it come from?
2) Which generator of `G1` (denoted by `g`) are we using?
3) Is `b2 == FQ2[4, 4]` (== 4 + 4 * i)? If so, what is the role of `b2`, where does it come from, and why is it called `b2`?
4) Could you add types for the arguments and return values of `hash_to_G2` and `modular_squareroot`?

File: specs/bls_verify.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (hwwhww) [reply specs/bls_verify.md]: Ah @vbuterin said we can actually remove this line. Shall we?

[the remark this answers is not in the document]
Comment (JustinDrake) [reply specs/bls_verify.md]: Done 👍

File: specs/bls_verify.md [hunk 4]
```diff
@@ -1,64 +1,121 @@
...
 
 ```python
-qmod = field_modulus ** 2 - 1
+qmod = q ** 2 - 1
 eighth_roots_of_unity = [FQ2([1,1]) ** ((qmod * k) // 8) for k in range(8)]
 
-def mod_sqrt(val):
-    candidate_sqrt = val ** ((qmod + 8) // 16)
-    check = candidate_sqrt ** 2 / val
+def modular_squareroot(value):
```
Comment (djrtwo) [inline specs/bls_verify.md]: Add type hint for args and return value

File: specs/bls_verify.md [hunk 5]
```diff
@@ -1,64 +1,121 @@
...
 G2_cofactor = 305502333931268344200999753193121504214466019254188142667664032982267604182971884026507427359259977847832272839041616661285803823378372096355777062779109
-field_modulus = 4002409555221667393417789825735904156556882819939007885332058136124031650490837864442687629129015664037894272559787
-
-def hash_to_G2(m, domain):
-    x1 = hash(bytes8(domain) + b'\x01' + m)
-    x2 = hash(bytes8(domain) + b'\x02' + m)
-    x_coord = FQ2([x1, x2]) # x1 + x2 * i
+q = 0x1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab
+
+def hash_to_G2(message, domain):
```
Comment (djrtwo) [inline specs/bls_verify.md]: Add typehints for args and return value

File: specs/bls_verify.md (hunk 4 above)
Comment (djrtwo) [reply specs/bls_verify.md]: Just saw the request for @vbuterin

**Comment by hwwhww** [comment]
Reference code: https://github.com/hwwhww/py-evm/blob/54cccde23d058a01ae68167f14d8f308a57140a2/eth/utils/bls.py

File: specs/bls_verify.md (hunk 5 above)
Comment (JustinDrake) [reply specs/bls_verify.md]: Done :)

File: specs/bls_verify.md (hunk 4 above)
Comment (JustinDrake) [reply specs/bls_verify.md]: Done

**Comment by JustinDrake** [comment]
I think it's good enough for a first pass cleanup :) I'll let you approve and merge @djrtwo.

**Pull request merged**